### ISSUE TITLE
Training Instability in COCONUT Latent Reasoning Stage

### ISSUE THREAD

**Issue by gdgdandsz** [facebookresearch/coconut#50]
I am attempting to reproduce the COCONUT latent reasoning training on the Qwen3-4B-Instruct model using the provided repository code. However, I observed significant instability and a decreasing trend in the testing accuracy as the number of latent tokens increased across epochs/stages. Is this a normal phenomenon?
Epoch | Latent Tokens | Test Accuracy
------|---------------|--------------
1     | 0             | 72%
2     | 2             | 48%
3     | 2             | 65%
4     | 4             | 68%
5     | 4             | 57%
6     | 6             | 65%
7     | 6             | 59%

**Comment by hzy051302** [comment]
When I replicated CODI, I also observed a similar phenomenon: the best result was achieved when the latent token was set to 0.
For CODI, I executed it according to the readme file. However, I found that the weights file provided by huggingface couldn't be used [facepalm]. Then, I used the weights of a GPT2 model that was trained for ten rounds on a 4090 GPU before. I increased the number of thinking times. acc: 28 ... - 25.85 - 26.84 - 27.07 - 27.37 - 27.22 - 27.22 - 27.29 - 27.14 - 27.07 - 26.84 - 26.69 - 26.76 - - - 26.67 - 26.91 - 26.91 - 26.91 - 26.84 - 26.46 - 26.54 - 26.38

**Comment by J1mL1** [comment]
Hi @gdgdandsz, may I ask what your hyperparameters are? I cannot even replicate your results. My result on Qwen3-4B-Instruct was shown below:

| Epoch | Latent Tokens | Test Accuracy |
|-------|---------------|---------------|
| 2     | 2             | 0.6944        |
| 3     | 2             | 0.6587        |
| 4     | 4             | 0.4484        |
| 5     | 4             | 0.5456        |
| 6     | 6             | 0.3810        |
| 7     | 6             | 0.4880        |
